controller/service.go

```diff
@@ -157,11 +157,15 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 
 	// At this point, we have an IP selected somehow, all that remains
 	// is to program the data plane.
-	lbIngressIPs := []v1.LoadBalancerIngress{}
-	for _, lbIP := range lbIPs {
-		lbIngressIPs = append(lbIngressIPs, v1.LoadBalancerIngress{IP: lbIP.String()})
+	lenLB := len(svc.Status.LoadBalancer.Ingress)
+	for i, lbIP := range lbIPs {
+		if i >= lenLB {
```

Comment:
can you explain the reasoning behind this logic? why do we append only if the lenght is bigger? What happens if the service is processed twice?

I think the issue raised is about the hostname part of an ingress, which in case of loadbalancer services must be still associated with an ip provided by metallb.

Reply:
Since the main idea is to only modify the IP, the cleaning/adding part will only remove/add an ingress when the other fields (hostname and ports) are unset. If there exist an ingress with a hostname for example, the IP will be overwritten.

To replicate the issue 1733, I created a controller that adds an ingress with a hostname value. When metallb controller assigns the IP, it will use the same ingress. The IP will be set. The intention is that both, hostname and IP, are in the same ingress. They are related.

When the service is processed twice, the IP will be added in the same place.

So, when there exist ingresses, the controller will set the IP on them. Otherwise, it will append a new ingress with the IP.

That's how I understood the request from the issue 1733, am I missing something?

Reply:
Ok, now I get it. What I find hard-ish to follow is that we are trying to recycle the current set of ips, tweaking the lenghts.
What I think it'd be a bit more straightful, is to use the newly produced values as the entry points and consume the hostnames added to the status part of the service.

So, what I think it'd be a bit more straightforward is to build a map out of the service we get from the k8s api, that relates each ip to its hostname, and use it to feed the hostname part of the ingress ip.
WDYT?

Reply:
I'm not tweaking the length. I just find out what the length of the array is, so that I don't index out of range. If it is possible to safely index the array, then the ip is set. Otherwise, a new ingress is created. I think that using a map will require more code. Not sure if it is simpler to understand.